### ISSUE TITLE
Resources rebuilt unncessarily during install or on subsequent builds.

### ISSUE THREAD

**Issue by sergey-239** [CauldronDevelopmentLLC/cbang#95]
Dear Joseph,
for some reason, the install target triggers rebuild of build/resources.os, which in turn makes the default target out of date.
Could you please change this behavior, so when there is no actual need to rebuild the default target (it was just successfully built), it is not linked again by 'scons install' invocation.

**Comment by jcoffland** [comment]
Does my recent commit fix this?

**Comment by sergey-239** [comment]
Joseph,

unfortunately, no. Building of build/resources.os is still performed in install target, so the default targets get rebuilt.
Thank you,
Sergey

**Comment by sergey-239** [comment]
Joseph.
after some digging into the problem I found that your patch is actually prevents the recreation of generated files. Though I would use more explicit test for install target like `'install' in BUILD_TARGETS`. Nevertheless scons decides for some reason that the `build/resources.os` deps are changed and compiles it. 
The solution is to invoke `scons install` with `--implicit-deps-unchanged` option switch: then it does not compile/build anything even without this commit.

**Comment by sergey-239** [comment]
for the record I am using python3-scons-4.4.0-2.fc38.noarch

**Comment by jcoffland** [comment]
I'm not sure why ``--implicit-deps-unchanged`` is necessary for your build.  It must have something to do with your particular configuration.  When I run ``scons install``.  No files are rebuilt.  The default build is a static build, maybe that has something to do with it.

**Comment by sergey-239** [comment]
Joseph,
I don't know either. I have a) very basic knowledge of python and b) I need to study the scons source to get to the roots of the problem. Unfortunately I have no time for this now. Presently I am building shared library, but I was encountering the same effect when I was building camotics with static cbang during install build of the camotics. Now I have a workaround, Probably would look into this issue some day.
Regards,
Sergey

**Comment by sergey-239** [comment]
Joseph, I have the hypotesy.
During default target build, scons prints out the following from resource tool emitter:
`('build/resources.cpp', 'src/resources.data/data2.cpp')`
During install target build, scons outputs this:
`('build/resources.cpp', 'build/resources.data/data2.cpp')`
This essentially means that the the nodes graph is different between two scons invocations, and scons has no cached time/checksum for node `build/resources.data/data2.cpp` as a dependency of 'build/resources.cpp', so it marks the latter as need to rebuild node.

Then, the next thing (just after the default target build):
```
ls -l --time-style=+%s build/resources.cpp src/resources.data/data2.cpp build/resources.data/data2.cpp
ls: cannot access 'src/resources.data/data2.cpp': No such file or directory
-rw-r--r--. 1 mockbuild mock  686 1663962660 build/resources.cpp
-rw-r--r--. 1 mockbuild mock 4469 1663962660 build/resources.data/data2.cpp
```
this explains everything, but the fact that you do not observe the similar behavior.

Unfortunately, I am not capable to fix this properly in a reasonable time: despite of my thirty five years experience in sw engineering, the functional languages are above my procedural way of thinking...

Regards,
Sergey

**Comment by jcoffland** [comment]
Yeah ``src/resources.data/data2.cpp`` does not and should not exist.  I'm not sure how that got in there.  When I build with your command line it prints this:

    ('build/resources.cpp', 'build/resources.data/data2.cpp', 'build/resources.data/data3.cpp', 'build/resources.data/data4.cpp', 'build/resources.data/data5.cpp', 'build/resources.data/data6.cpp', 'build/resources.data/data7.cpp', 'build/resources.data/data8.cpp', 'build/resources.data/data9.cpp', 'build/resources.data/data10.cpp', 'build/resources.data/data11.cpp')

Did you remove the license files from ``src/resources/licenses/``?  Maybe I should add an option to just disable the compiled in resources.  The only reason they are there in cbang is to be able to display these license files.

**Comment by sergey-239** [comment]
Ah, yes, I forgot to mention that any third-party code is removed in the %prep phase, so there is only two subdirectories under src/: cbang and resources. also, all licenses from /resources/licensed removed except cbang.txt. This is to ensure that no third-party code is used in build

**Comment by sergey-239** [comment]
By the way, even when the source code is not removed, but a library disabled by disable_local, its corresponding license should not be included into resources.os, do you agree?

**Comment by sergey-239** [comment]
> Did you remove the license files from `src/resources/licenses/`? Maybe I should add an option to just disable the compiled in resources. The only reason they are there in cbang is to be able to display these license files.

Yes, I think that if lib source is disabled by disable_local, then appropriate license should not be included as well. The reason to include license texts at all is obvious.

**Comment by sergey-239** [comment]
Two more things to start the install target build using commands I posted above:
1) You need to create a ./lib/libcbang0.so symlink to /.lib/cbang0.so.1.7.0
2) You need to export CBANG_LIBNAME=cbang0 
to let scons find the built library

**Comment by sergey-239** [comment]
Joseph,

#99 does not fix the issue with wrong deps list:
default target:
`('build/resources.cpp', 'src/resources.data/data2.cpp', 'src/resources.data/data3.cpp')`
install target:
`('build/resources.cpp', 'build/resources.data/data2.cpp', 'build/resources.data/data3.cpp')`

Regards,
Sergey

**Comment by sergey-239** [comment]
Joseph,

The problem's source is identified.
Unless the `build/resources.data/data2.cpp` exists, the `File('%s/data%d.cpp' % (data_dir, id))` at [config/resources/__init__:](https://github.com/CauldronDevelopmentLLC/cbang/blob/3cf77de61025e9652450e8b5ef1b301a75ff5305/config/resources/__init__.py#L211) returns the object with path modified to `src/...`. 
You do not observe the same behavior as you most likely start the default build from dirty source tree while mock does this always from scratch.
To simulate this behavior just remove `build/resources.data` before invoking scons.
Also, there is some strange thing with `build/resources.data/data2.cpp`'s dependencies:
a) with or without my patch #99 and some(all) disabled libs, the deps are all files existing in the src/resources/ subtree;
b) the deps are listed three times contrary to regular source files in `src/` subtree:
```
  +-build/resources.data/data2.os
  | +-build/resources.data/data2.cpp
  | | +-src/resources
  | | | +-src/resources/licenses
  | | | | +-src/resources/licenses/boost.txt
  | | | | +-src/resources/licenses/bzip2.txt
  | | | | +-src/resources/licenses/cbang.txt
  | | | | +-src/resources/licenses/expat.txt
  | | | | +-src/resources/licenses/libevent.txt
  | | | | +-src/resources/licenses/libyaml.txt
  | | | | +-src/resources/licenses/lz4.txt
  | | | | +-src/resources/licenses/re2.txt
  | | | | +-src/resources/licenses/sqlite.txt
  | | | | +-src/resources/licenses/test.backup.file~
  | | | | +-src/resources/licenses/zlib.txt
  | | | +-src/resources/test.backup.file~
  | | +-src/cbang/util/Resource.h
  | | +-src/resources/licenses
  | | | +-src/resources/licenses/boost.txt
  | | | +-src/resources/licenses/bzip2.txt
  | | | +-src/resources/licenses/cbang.txt
  | | | +-src/resources/licenses/expat.txt
  | | | +-src/resources/licenses/libevent.txt
  | | | +-src/resources/licenses/libyaml.txt
  | | | +-src/resources/licenses/lz4.txt
  | | | +-src/resources/licenses/re2.txt
  | | | +-src/resources/licenses/sqlite.txt
  | | | +-src/resources/licenses/test.backup.file~
  | | | +-src/resources/licenses/zlib.txt
  | | +-src/resources/test.backup.file~
  | | +-src/resources/licenses/boost.txt
  | | +-src/resources/licenses/bzip2.txt
  | | +-src/resources/licenses/cbang.txt
  | | +-src/resources/licenses/expat.txt
  | | +-src/resources/licenses/libevent.txt
  | | +-src/resources/licenses/libyaml.txt
  | | +-src/resources/licenses/lz4.txt
  | | +-src/resources/licenses/re2.txt
  | | +-src/resources/licenses/sqlite.txt
  | | +-src/resources/licenses/test.backup.file~
  | | +-src/resources/licenses/zlib.txt
  | +-src/cbang/util/Resource.h
  
```
Regards,
Sergey

**Comment by sergey-239** [comment]
Joseph, 

it looks like the list of dependencies includes whatever is in the supplied source directory as you use simple directory scanner in defining resources builder. I understand that this does not influences outcome, it just could trigger unnecessary builds when something get touched in the source directory, that is not my case anyway. 
As I mentioned already, I have the workaround so this is not an urgent thing to fix.
At least this issue enforced me to start study scons: I have never seen it before. To be honest, I like scons more and more and as a side effect I will improve my python skills by exploring scons ;)

Regards,
Sergey

**Comment by jcoffland** [comment]
This seems to be the exact same problem: https://jira.mongodb.org/browse/SERVER-33111  It looks like they fixed it by patching scons but the fix was apparently not applied upstream.

**Comment by jcoffland** [comment]
I've created a simplified test case which exhibits the bug.

[scons-v4.0.1-generated-file-rebuild-on-install-bug-20220926.zip](https://github.com/CauldronDevelopmentLLC/cbang/files/9646143/scons-v4.0.1-generated-file-rebuild-on-install-bug-20220926.zip)

Running ``scons`` I get:

```
scons: Reading SConscript files ...
New targets ('build/resources.cpp', 'src/resources.data/hello.txt.cpp')
scons: done reading SConscript files.
scons: Building targets ...
build(["build/resources.cpp", "build/resources.data/hello.txt.cpp"], ["src/resources"])
Writing build/resources.cpp
Writing build/resources.data/hello.txt.cpp
g++ -o build/resources.data/hello.txt.o -c -Isrc build/resources.data/hello.txt.cpp
g++ -o build/resources.o -c -Isrc build/resources.cpp
g++ -o test build/resources.o build/resources.data/hello.txt.o
Install file: "test" as "tmp/test"
scons: done building targets.
```

Then run ``scons --debug=explain install``:

```
scons: Reading SConscript files ...
New targets ('build/resources.cpp', 'build/resources.data/hello.txt.cpp')
scons: done reading SConscript files.
scons: Building targets ...
scons: rebuilding `build/resources.o' because `src/dep/dep.h' is no longer a dependency
g++ -o build/resources.o -c -Isrc build/resources.cpp
scons: `install' is up to date.
scons: done building targets.
```

It rebuilds every time I switch from ``scons install`` to ``scons`` and back.  Repeated runs of the same target do not rebuild.

It appears that running ``scons`` vs ``scons install`` writes the dependencies to the scons database differently.

**Comment by sergey-239** [comment]
Joseph
thank you for the references.

> It appears that running `scons` vs `scons install` writes the dependencies to the scons database differently.

No, not only in this situation. Try this: 
```
rm -rf build/resources.*
scons --debug=explain
scons --debug=explain
scons --debug=explain

```
On first invocation of scons it rebuilds resources as there are no cpp files
On second - because of scons bug with dependencies
And only on third it says that target is up to date. However, it rebuilds build_info as marked 'always build' but then this does not make default target out of date as checksum of build_info.os remains the same, obviously.

By the way, I completed the #99. Now it does not include unnecessary licenses and also builds proper dependencies lists, so the mod time check inside resources_build no longer needed.
As for scons bug with deps, I will continue to find if this behavior might be avoided without patching scons

Regards,
Sergey

**Comment by jcoffland** [comment]
It will rebuild on each invocation like this too:

```
rm -rf build/resources.*
scons
scons
scons  install
scons
```

**Comment by sergey-239** [comment]
> It will rebuild on each invocation like this too:
> 
> ```
> rm -rf build/resources.*
> scons
> scons
> scons  install
> scons
> ```

Yes, I did not check all possible combinations :)
I checked the scons source installed on my system form Fedora repo: it has at least some changes from https://github.com/SCons/scons/pull/3340 incorporated.
Regards,
Sergey

**Comment by jcoffland** [comment]
My scons also has that "fix" applied.  It seems that scons has more problems with tracking dependencies correctly when a builder adds additional targets.  Thanks for helping track this down.  It would be great if we could find a solution that does not require modifying scons but you work around is ok for your situation and the extra rebuilding does not bother me too much so it's probably not a high priority issue.

**Comment by jcoffland** [comment]
I've reopened this issue to track the bug even though the true but is upstream in scons.
